diff --git a/CMakeLists.txt b/CMakeLists.txt
index c6ee47cd9..8514ff119 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1294,6 +1294,10 @@ if (MINGW)
 	add_definitions(-DWINVER=0x0601 -D_WIN32_WINNT=0x0601)
 endif()
 
+if (TERMUX)
+  add_definitions(-DTERMUX_PREFIX=\"/data/data/com.termux/files/usr\")
+endif()
+
 if (HDR_PRIVATE)
 	source_group("Headers Private"  FILES ${HDR_PRIVATE})
 endif()
diff --git a/lib/plat/unix/unix-resolv.c b/lib/plat/unix/unix-resolv.c
index 63579c6a6..5080b68f8 100644
--- a/lib/plat/unix/unix-resolv.c
+++ b/lib/plat/unix/unix-resolv.c
@@ -25,6 +25,12 @@
 #include "private-lib-core.h"
 #include "private-lib-async-dns.h"
 
+#ifndef TERMUX_PREFIX
+#define PREFIX ""
+#else
+#define PREFIX TERMUX_PREFIX
+#endif
+
 /*
  * Shared "nameserver" line parser for /etc/resolv.conf-shaped files.
  *
@@ -59,7 +65,7 @@ lws_asyncdns_parse_resolv_conf(struct lws_context *context, int index,
 	if (!r)
 		return -1;
 
-	fd = open(env && env[0] ? env : "/etc/resolv.conf", LWS_O_RDONLY);
+	fd = open(env && env[0] ? env : PREFIX "/etc/resolv.conf", LWS_O_RDONLY);
 	if (fd < 0)
 		goto bail;
 
