diff --git a/lws-context.c b/lws-context.c
index ae8ad8d..7d556b6 100644
--- a/lws-context.c
+++ b/lws-context.c
@@ -13,6 +13,7 @@
 #include "lws.h"
 #include "js-utils.h"
 #include "iohandler.h"
+#include "lws-epoll.h"
 
 #define LWS_PLUGIN_STATIC
 
@@ -776,6 +777,8 @@ context_new(JSContext* ctx) {
 
 static void
 context_free(JSRuntime* rt, LWSContext* lc) {
+  lws_epoll_destroy(lc);
+
   if(lc->js) {
     JS_FreeContext(lc->js);
     lc->js = NULL;
@@ -864,6 +867,7 @@ lwsjs_context_methods(JSContext* ctx, JSValueConst this_val, int argc, JSValueCo
   switch(magic) {
     case METHOD_DESTROY: {
       if(lc->ctx) {
+        lws_epoll_destroy(lc);
         lws_context_destroy(lc->ctx);
         lc->ctx = NULL;
         ret = JS_TRUE;
@@ -1140,8 +1144,7 @@ callback_pollfd(struct lws* wsi, enum lws_callback_reasons reason, void* user, v
     case LWS_CALLBACK_DEL_POLL_FD: {
       struct lws_pollargs* x = in;
 
-      iohandler_set(lc, x->fd, JS_NULL, 0);
-      iohandler_set(lc, x->fd, JS_NULL, 1);
+      lws_epoll_del(lc, x->fd);
       return 0;
     }
 
@@ -1152,21 +1155,7 @@ callback_pollfd(struct lws* wsi, enum lws_callback_reasons reason, void* user, v
       if(x->events == x->prev_events)
         return 0;
 
-      BOOL write = !!(x->events & POLLOUT);
-      JSValueConst data[] = {
-          JS_NewInt32(ctx, x->fd),
-          JS_NewInt32(ctx, x->events),
-          JS_NewBool(ctx, write),
-          JS_NewInt64(ctx, (intptr_t)lws_get_context(wsi)),
-      };
-      JSValue fn = JS_NewCFunctionData(ctx, protocol_handler, 0, 0, countof(data), data);
-
-      if(reason == LWS_CALLBACK_CHANGE_MODE_POLL_FD)
-        iohandler_set(lc, x->fd, JS_NULL, !write);
-
-      iohandler_set(lc, x->fd, fn, write);
-
-      JS_FreeValue(ctx, fn);
+      lws_epoll_ctl(lc, x->fd, x->events);
       return 0;
     }
 
diff --git a/lws-context.h b/lws-context.h
index 6095f93..81f37d5 100644
--- a/lws-context.h
+++ b/lws-context.h
@@ -7,11 +7,14 @@
 #include <libwebsockets.h>
 #include "lws.h"
 
+typedef struct LWSEpoll LWSEpoll;
+
 typedef struct LWSContext {
   struct lws_context* ctx;
   struct lws_context_creation_info info;
   JSContext* js;
   struct list_head handlers;
+  LWSEpoll* epoll;
 } LWSContext;
 
 typedef struct {
